packages/eslint-plugin/src/rules/prefer-regexp-exec.ts

```diff
@@ -97,38 +128,46 @@ export default createRule({
           });
         }
 
-        if (isRegExpType(argumentNode)) {
-          return context.report({
-            node: memberNode.property,
-            messageId: 'regExpExecOverStringMatch',
-            fix: getWrappingFixer({
-              sourceCode,
-              node: callNode,
-              innerNode: [objectNode, argumentNode],
-              wrap: (objectCode, argumentCode) =>
-                `${argumentCode}.exec(${objectCode})`,
-            }),
-          });
-        }
+        const argumentType = typeChecker.getTypeAtLocation(
+          parserServices.esTreeNodeToTSNodeMap.get(argumentNode),
+        );
+        const argumentTypes = collectArgumentTypes(
+          tsutils.unionTypeParts(argumentType),
+        );
+        switch (argumentTypes) {
+          case ArgumentType.Both:
+          case ArgumentType.Other:
+            return context.report({
+              node: memberNode.property,
+              messageId: 'regExpExecOverStringMatch',
+            });
```

Comment:
Are we sure we want to report if there is both string and regex?
It might be a bit weird for a user as they would have to create forked code in order to satisfy the rule.

If we do want to report - might be worth putting it behind a flag maybe?

Reply:
I think I'd just gotten confused originally 😄  agreed. This no longer reports if the type isn't purely a regexp or string.